File: packages/cli/cobra/src/services/status/status.go
package status

import (
	"encoding/json"
	"fmt"
	"log"

	"time"

	"github.com/hieudoanm/hieudoanm/src/libs/requests"
)

// Page represents the status page info
type Page struct {
	Id        string `json:"id"`
	Name      string `json:"name"`
	Url       string `json:"url"`
	TimeZone  string `json:"time_zone"`
	UpdatedAt string `json:"updated_at"`
}

// Status represents the page status
type Status struct {
	Indicator   string `json:"indicator"`
	Description string `json:"description"`
}

// Response wraps page and status
type Response struct {
	Page   Page   `json:"page"`
	Status Status `json:"status"`
}

// GetStatus fetches status from a URL
func GetStatus(url string, debug bool) (Response, error) {
	responseByte, getError := requests.Get(url, requests.Options{Debug: debug})
	if getError != nil {
		return Response{}, getError
	}
	var response Response
	jsonError := json.Unmarshal(responseByte, &response)
	if jsonError != nil {
		return Response{}, jsonError
	}
	return response, nil
}

// PrintFullStatus prints status with borders, color, and timestamp
func PrintFullStatus(url string, debug bool) {
	resp, err := GetStatus(url, debug)
	if err != nil {
		log.Printf("[%s] \033[31mError:\033[0m %v\n", time.Now().Format(time.RFC3339), err)
		return
	}

	// Colors
	const (
		ColorReset  = "\033[0m"
		ColorBlue   = "\033[34m"
		ColorGreen  = "\033[32m"
		ColorYellow = "\033[33m"
		ColorCyan   = "\033[36m"
	)

	timestamp := time.Now().Format(time.RFC3339)
	border := fmt.Sprintf("%s==================== STATUS PAGE ==================== [%s]%s", ColorBlue, timestamp, ColorReset)

	fmt.Println(border)
	fmt.Printf("%sPage Name    :%s %s\n", ColorCyan, ColorReset, resp.Page.Name)
	fmt.Printf("%sPage ID      :%s %s\n", ColorCyan, ColorReset, resp.Page.Id)
	fmt.Printf("%sURL          :%s %s\n", ColorCyan, ColorReset, resp.Page.Url)
	fmt.Printf("%sTime Zone    :%s %s\n", ColorCyan, ColorReset, resp.Page.TimeZone)
	fmt.Printf("%sUpdated At   :%s %s\n", ColorCyan, ColorReset, resp.Page.UpdatedAt)
	fmt.Printf("%sIndicator    :%s %s\n", ColorGreen, ColorReset, resp.Status.Indicator)
	fmt.Printf("%sDescription  :%s %s\n", ColorYellow, ColorReset, resp.Status.Description)
	fmt.Println(border)
}

// GetDescriptiveStatus
func GetDescriptiveStatus(name string, url string, debug bool) string {
	resp, err := GetStatus(url, debug)
	if err != nil {
		return fmt.Sprintf("\033[31mError:\033[0m %v", err)
	}

	if resp.Status.Indicator == "none" {
		return "✅ Healthy"
	}

	return "❌ Offline"
}

// PrintDescriptiveStatus prints only the descriptive status with timestamp
func PrintDescriptiveStatus(name string, url string, debug bool) {
	resp, err := GetStatus(url, debug)
	if err != nil {
		log.Printf("[%s] \033[31mError:\033[0m %v\n", time.Now().Format(time.RFC3339), err)
		return
	}
	const (
		ColorYellow = "\033[33m"
		ColorReset  = "\033[0m"
	)
	timestamp := time.Now().Format(time.RFC3339)
	fmt.Printf("[%s] %s%s : %s%s\n", timestamp, ColorYellow, name, resp.Status.Description, ColorReset)
}

File: packages/cli/hieudoanm/rust/src/services/status.rs
use anyhow::{Context, Result};
use chrono::Utc;
use serde::Deserialize;

#[derive(Debug, Deserialize)]
pub struct Page {
    pub id: String,
    pub name: String,
    pub url: String,
    pub time_zone: String,
    pub updated_at: String,
}

#[derive(Debug, Deserialize)]
pub struct Status {
    pub indicator: String,
    pub description: String,
}

#[derive(Debug, Deserialize)]
pub struct Response {
    pub page: Page,
    pub status: Status,
}

pub fn get_status(url: &str) -> Result<Response> {
    let resp = reqwest::blocking::get(url).context("failed to fetch status page")?;
    let response: Response = resp.json().context("failed to parse status response")?;
    Ok(response)
}

pub fn print_full_status(url: &str) {
    match get_status(url) {
        Ok(resp) => {
            let timestamp = Utc::now().to_rfc3339();
            let border = format!(
                "==================== STATUS PAGE ==================== [{}]",
                timestamp
            );
            println!("{}", border);
            println!("Page Name    : {}", resp.page.name);
            println!("Page ID      : {}", resp.page.id);
            println!("URL          : {}", resp.page.url);
            println!("Time Zone    : {}", resp.page.time_zone);
            println!("Updated At   : {}", resp.page.updated_at);
            println!("Indicator    : {}", resp.status.indicator);
            println!("Description  : {}", resp.status.description);
            println!("{}", border);
        }
        Err(e) => {
            eprintln!("[{}] Error: {}", Utc::now().to_rfc3339(), e);
        }
    }
}

pub fn get_descriptive_status(name: &str, url: &str) -> String {
    match get_status(url) {
        Ok(resp) => {
            if resp.status.indicator == "none" {
                format!("{} - Healthy", name)
            } else {
                format!("{} - Offline", name)
            }
        }
        Err(e) => format!("{} - Error: {}", name, e),
    }
}

pub fn print_descriptive_status(name: &str, url: &str) {
    let status = get_descriptive_status(name, url);
    let timestamp = Utc::now().to_rfc3339();
    println!("[{}] {}", timestamp, status);
}
